File: app/src/main/java/com/example/healthiconiq/CHATGPT_API.java
package com.example.healthiconiq;

import android.net.Uri;

import com.google.firebase.crashlytics.buildtools.reloc.org.apache.http.client.HttpClient;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;

public class CHATGPT_API {

    public static String getData(Uri imageUri, String language, String apiKey) {
        try {
            // Read the image file as bytes
            byte[] imageBytes = Files.readAllBytes(Paths.get(imageUri.getPath()));

            // Create the boundary
            String boundary = "-----011000010111000001101001";

            // Build the multipart form data body
            String body = "--" + boundary + "\r\n" +
                    "Content-Disposition: form-data; name=\"language\"\r\n\r\n" +
                    language + "\r\n" +
                    "--" + boundary + "\r\n" +
                    "Content-Disposition: form-data; name=\"api_key\"\r\n\r\n" +
                    apiKey + "\r\n" +
                    "--" + boundary + "\r\n" +
                    "Content-Disposition: form-data; name=\"image\"; filename=\"health.png\"\r\n" +
                    "Content-Type: image/png\r\n\r\n" +
                    new String(imageBytes) + "\r\n" +
                    "--" + boundary + "--\r\n";

            // Create the HTTP request
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create("https://medical.pythonanywhere.com/process_image"))
                    .header("Accept", "*/*")
                    .header("User-Agent", "Thunder Client (https://www.thunderclient.com)")
                    .header("Content-Type", "multipart/form-data; boundary=" + boundary)
                    .POST(HttpRequest.BodyPublishers.ofString(body))
                    .build();

            // Send the request and get the response
            HttpResponse<String> response = HttpClient.newHttpClient().send(request, HttpResponse.BodyHandlers.ofString());

            // Return the response body
            return response.body();
        } catch (Exception e) {
            e.printStackTrace();
            return "Error: " + e.getMessage();
        }
    }
}

File: app/src/main/java/com/example/healthiconiq/CHATGPT_API.kt
import android.content.Context
import android.net.Uri
import android.util.Log
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import okhttp3.logging.HttpLoggingInterceptor

interface DataCallback {
    fun onSuccess(data: String)
    fun onFailure(error: String)
}

object CHATGPT_API {
    private val client = OkHttpClient()

    fun getData(context: Context, imageUri: Uri, language: String, apiKey: String, callback: DataCallback) {
        val prompt = "Consider this medical symbol image and explain it or give me the meaning of it with in short limit of one sentence in: $language"
        Log.d("API key in getData: ",apiKey)
        context.contentResolver.openInputStream(imageUri)?.use { inputStream ->
            val bytes = inputStream.readBytes()
            val requestBody = bytes.toRequestBody("image/png".toMediaTypeOrNull(), 0)

            val multipartBody = MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("prompt", prompt)
                .addFormDataPart("language", language)
                .addFormDataPart("api_key", apiKey)
                .addFormDataPart("image", "image.png", requestBody)
                .build()

            val logging = HttpLoggingInterceptor().apply {
                level = HttpLoggingInterceptor.Level.BODY
            }
            val client = OkHttpClient.Builder()
                .addInterceptor(logging)
                .build()

            val request = Request.Builder()
                .url("https://medical.pythonanywhere.com/process_image")
                .post(multipartBody)
                .build()

            println("Request details: ${request.headers}")

            client.newCall(request).enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    callback.onFailure("Error on API Call: ${e.message}")
                }

                override fun onResponse(call: Call, response: Response) {
                    response.use { resp ->
                        val responseBody = resp.body?.string()
                        if (responseBody != null) {
                            callback.onSuccess(responseBody)
                        } else {
                            callback.onFailure("Received empty response")
                        }
                    }
                }
            })
        } ?: run {
            println("Error: Unable to open image stream.")
            callback.onFailure("Failed to initiate request due to input stream error.")
        }
    }
}
